core/modules/entity/entity.controller.inc

```diff
@@ -142,14 +160,42 @@ class DefaultEntityController implements EntityControllerInterface {
     }
 
     // Check if the entity type supports static caching of loaded entities.
-    $this->cache = !empty($this->entityInfo['static cache']);
+    $this->staticCache = !empty($this->entityInfo['static cache']);
+
+    // Check if the entity type supports usage of the persistant entity cache.
+    if (isset($this->entityInfo['entity cache'])
+      && $this->entityInfo['entity cache']
+      && db_table_exists('cache_entity_' . $this->entityType)) {
+      $this->persistanctCache = TRUE;
+    }
+    else {
+      $this->persistanctCache = FALSE;
+    }
   }
 
   /**
    * Implements EntityControllerInterface::resetCache().
    */
   public function resetCache(array $ids = NULL) {
-    if (isset($ids)) {
+    // Reset the persistent cache.
+    if ($this->persistanctCache) {
+      if (!empty($ids)) {
+        cache_clear_all($ids, 'cache_entity_' . $this->entityType);
+      }
+      else {
+        // Force all cached entries to be deleted.
+        cache_flush('cache_entity_' . $this->entityType);
+      }
+    }
+
+    $this->resetStaticCache($ids);
+  }
+
+  /**
+   * Implements EntityControllerInterface::resetStaticCache().
+   */
+  public function resetStaticCache(array $ids = NULL) {
+    if (isset($ids) && !empty($ids)) {
```

Comment:
Minor, but trying to think of why `!empty()` wouldnt suffice, without the `isset`.

Reply:
I could see this being helpful in the event you were going to call resetStaticCache() with a list of IDs but you weren't sure if there was going to be something in the array. For example this may or may not have values in the array:

```
$this->resetStaticCache($unknown_array);
```

If it contained one ID, it would clear one item from the cache. If it was an empty array, it shouldn't clear anything at all. Versus if NULL is passed in or the parameter omitted, everything is cleared.

Reply:
I through the current code was good initially but after writing the above, it seems like only `isset()` should be used, and remove the `!empty()`. Because an empty array shouldn't clear anything. Only omitting the array entirely or passing in NULL should clear the entire cache.